File: atcoder/abc/abc123/d.py
X, Y, X, K = map(int, input().split())
A = list(map(int, input().split()))
B = list(map(int, input().split()))
C = list(map(int, input().split()))
A.sort(reverse=True)
B.sort(reverse=True)
C.sort(reverse=True)

ab = [a+b for a in A for b in B]
ab.sort(reverse=True)

ans = [x+c for x in ab[:K] for c in C]
ans.sort(reverse=True)
print('\n'.join(map(str, ans[:K])))

File: atcoder/abc/abc123/d.rb
*_, k = gets.split.map(&:to_i)
a = gets.split.map(&:to_i).sort.reverse
b = gets.split.map(&:to_i).sort.reverse
c = gets.split.map(&:to_i).sort.reverse
puts a.product(b).map { |x, y| x + y }.sort.reverse[0, k].product(c).map{ |x, y| x + y }.sort.reverse[0, k]
# require "benchmark"
# arr = [*0...20000000].shuffle
# Benchmark.bm(12) do |x|
#   x.report("sort".ljust(12)) { arr.sort { |x, y| y <=> x } }
#   x.report("sort_by".ljust(12)) { arr.sort_by { |x| -x } }
#   x.report("sort.reverse".ljust(12)) { arr.sort.reverse }
# end
